scipy/_lib/_array_api.py

```diff
@@ -191,33 +192,81 @@ def is_torch(xp):
     return xp.__name__ == 'scipy._lib.array_api_compat.array_api_compat.torch'
 
 
-def assert_equal(actual, desired, err_msg='', xp=None):
+def _strict_check(actual, desired, check_xp=True, check_dtype=True, check_shape=True):
+    try:
+        actual.dtype
+        desired.dtype
+    except AttributeError:
+        raise ValueError("Inputs should be arrays with a dtype attribute, "
+                         "python scalars and arrays are not accepted.")
+    if check_xp:
+        _assert_matching_namespace(actual, desired)
+    if check_dtype:
+        assert_(actual.dtype == desired.dtype,
+                f"Desired dtype: {desired.dtype}, actual dtype: {actual.dtype}")
+    if check_shape:
+        msg = (f"Array shapes are not equal: actual shape = {actual.shape}, "
+               f"desired shape = {desired.shape}")
+        assert_(actual.shape == desired.shape, msg)
+
+
+def _assert_matching_namespace(actual, desired):
+    desired_space = array_namespace(desired)
+    if isinstance(actual, tuple):
```

Comment:
Is `_assert_matching_namespace` going to be used elswhere with tuple `actual`? The way it's used here, it looks like it will always have a `dtype` attribute.

If so, I'd suggest `actual = (actual,) if not isinstance(actual, tuple) else actual` or similar. Then you don't need to copy-paste the logic. Besides the simplification, it would avoid inconsistencies - the assertions are written differently, and one has a message while the other doesn't. (Or was this intentional?)

Reply:
I don't think that it is currently used with tuple `actual` anywhere, I'm not sure about the future.

> actual = actual if isinstance(actual, tuple) else (actual,)

Sounds good.

Reply:
> the assertions are written differently, and one has a message while the other doesn't. (Or was this intentional?)

Oops, that is a bug, the custom message should apply for both.

Reply:
I've addressed this in https://github.com/scipy/scipy/pull/19186/commits/2df7541e02f45415d8ecdc99b0e8f64a5af3043d. cc @tylerjereddy as original author.